Version vector prototype: propagate latest commit version from client to storage server

sbodagala opened the pull request:
Main changes:
- Propagate latest commit version from client to storage server
- Make a storage server read at the latest commit version (instead of the actual read version)
- Make a peek request block on a tlog until the tlog has a commit that is relevant to the requester (code changes to accomplish this have been extracted from https://github.com/apple/foundationdb/pull/5058)
- Propagate version vector deltas between processes

Changes related to the propagation of the latest commit version from client to storage server and to make a storage server read at the latest commit version:

CommitProxyInterface.h:
- Let the proxy return the tags of storage servers (along with the storage server interfaces) as part of GetKeyServerLocationsReply.

DatabaseContext.h:
- Map the tags of storage servers to their corresponding storage server interfaces.
- APIs to update the tag mapping and to fetch the tags of a given set storage server replicas.

NativeAPI.cpp:
- Update tag mappings in GetKeyServerLocationsReply() and getKeyRangeLocations_internal().
- Propagate the late commit versions of storage server replicas when making the following requests: 					                          GetValueRequest, GetKeyRequest, GetKeyValuesRequest, GetKeyValuesStreamRequest.

StorageServerInterface.h:
- Extend the following RPCs to also propagate the latest commit versions of all storage server replicas that serve the given key/key range: GetValueRequest, GetKeyRequest, GetKeyValuesRequest, GetKeyValuesStreamRequest.

VersionVector.h:
- Add APIs to set the versions for a given set of storage servers and to clear the vector.

CommitProxyServer.actor.cpp:
- Populate the storage server tag info in GetKeyServerLocationsReply.

GrvProxyServer.actor.cpp:
- Order the includes.

storageserver.actor.cpp:
- Make the server read at the specified latest commit version while serving the following types of requests: GetValueRequest, GetKeyRequest, GetKeyValuesRequest, GetKeyValuesStreamRequest.

Changes to make a peek request block on a tlog until the tlog has a commit that is relevant to the requester:

TLogServer.actor.cpp:
- Code changes extracted from https://github.com/apple/foundationdb/pull/5058

Changes related to the propagation of version vector deltas:

VersionVector.h:

- Add APIs to get and apply version vector deltas.

CommitProxyInterface.h:
- Extend GetReadVersionRequest and GetRawCommittedVersionRequest data structures so the client and the grv proxy can inform the maximum version in their version vectors.

NativeAPI.actor.cpp and GrvProxyServer.actor.cpp:
- Modify code so the client/grv proxy will include the maximum version in its version vector cache in GetReadVersionRequest/GetRawCommittedVersionRequest and then apply the version vector delta (that it receives from the grv proxy/sequenver) to its version vector cache.

CommitProxyServer.actor.cpp:
- Set "maxVersion = invalidVersion" in the GetReadVersionRequest that it makes.

masterserver.actor.cpp:
- Return version vector delta to the grv proxy.

# Code-Reviewer Section

The general guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `master` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

sbodagala commented on fdbclient/DatabaseContext.h:
```diff
@@ -434,13 +437,23 @@ class DatabaseContext : public ReferenceCounted<DatabaseContext>, public FastAll
+
+	// Returns the latest commit versions that mutated the specified storage servers
+	/// @note returns the latest commit version for a storage server only if the latest
+	// commit version of that storage server is below the specified "readVersion".
```
Does a client ever do reads at a version that is below the latest commit version in the database? If so, we won't need to do the filtering based on "readVersion".

dlambrig replied:
I thought a client is allowed to `set_read_version` ([link](https://apple.github.io/foundationdb/api-python.html#versions)) , so could choose a version below the last commit version.

sbodagala replied:
Ok, thanks for that info!

sbodagala commented on fdbclient/StorageServerInterface.h:
```diff
@@ -212,14 +213,23 @@ struct GetValueRequest : TimedRequest {
 	ReplyPromise<GetValueReply> reply;
+	VersionVector ssLatestCommitVersions; // includes the latest commit versions, as known
+	                                      // to this client, of all storage replicas that
+	                                      // serve the given key
```
Propagating the latest commit versions of all replicas here. This way the receiving storage server will get its latest commit version irrespective of the replica that the load balancer chooses to send the message to. Since the number of storage server replicas for a given shard is not high, this shouldn't add a significant amount of overhead. Let me know if it is preferable to modify the load balancer code over this.

Note: We will need to make sure that the latest commit versions of storage servers stay up-to-date (in the version vector on the sequencer) in the presence of key range splits/merges. Not addressing that issue in this PR.

dlambrig replied:
>Note: We will need to make sure that the latest commit versions of storage servers stay up-to-date (in the version vector on the sequencer) in the presence of key range splits/merges. Not addressing that issue in this PR.

Can you add a TODO to the code

sbodagala replied:
Done (placed that in masterserver.actor.cpp).

sbodagala commented on fdbclient/VersionVector.h:
```diff
@@ -43,6 +43,14 @@ struct VersionVector {
 		maxVersion = version;
 	}
 
+	void setVersion(const std::set<Tag>& tags, Version version) {
```
Extracted from Dan's PR.

dlambrig replied:
you can rebase as PR #5114 has merged. That may fix the conflict as well.

sbodagala replied:
Ok.

dlambrig commented on fdbserver/storageserver.actor.cpp:
```diff
@@ -1235,7 +1235,12 @@ ACTOR Future<Void> getValueQ(StorageServer* data, GetValueRequest req) {
-		state Version version = wait(waitForVersion(data, req.version, req.spanContext));
+		// If the client specified the latest commit version (that mutated the shard(s) being served
+		// by this storage server) then return the value that corresponds to that version.
+		Version readVersion = req.ssLatestCommitVersions.hasVersion(data->tag)
```
when would `ssLatestCommitVersions` not have `data->tag` ?

sbodagala replied:
> * when would `ssLatestCommitVersions` not have `data->tag` ?
I can think of the following scenarios: (a) if  the read is being done at an older version (the client won't propagate the latest commit version in that case, as it doesn't need to) (b) when the cluster/sequencer starts the version vector will be empty (and tags won't be added to the version vector until a transaction mutates them) (b) when a new storage server is added to the system? (I don't know if we allow new storage servers to be added to a running cluster but if we do then I don't think we will have entries for those until a transaction mutates them).

> * Consider making this `if` statement an inline function, as I see it used elsewhere.

Good suggestion. Done.

> * we now have a knob `ENABLE_VERSION_VECTOR`. We should probably not access things like the version vector when the knob is false.

Since we are working on a branch I am not sure we need to protect all the new code with the knob.  But maybe we should protect some new code with the knob (until we are ready to test this feature end-to-end) if we think that code would break simulation tests - if so maybe we should not populate the version vector on the sequencer (or protect that code with the knob) and then accessing the version vector here won't have any effect.

dlambrig replied:
When I debug failed simulation tests, I like to have a single switch to turn off the feature. This way I can easily run with and then without the feature enabled. 
>if so maybe we should not populate the version vector on the sequencer 

yes, I'd only put the switch around code that is impactful, i.e. not everywhere.

sbodagala replied:
> Ok, will put the switch around code that populates the version vector on the sequencer.

This part has been done already (by Dan):

```
if (SERVER_KNOBS->ENABLE_VERSION_VECTOR && req.writtenTags.present()) {
					// NB: this if-condition is not needed after wait-for-prev is ported to this branch
					if (req.version > self->ssVersionVector.maxVersion) {
						// TraceEvent("Received ReportRawCommittedVersionRequest").detail("Version",req.version);
						self->ssVersionVector.setVersion(req.writtenTags.get(), req.version);
					}


```

dlambrig replied:
[others](https://github.com/apple/foundationdb/pull/5182)

dlambrig commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -169,6 +169,34 @@ void DatabaseContext::removeTssMapping(StorageServerInterface const& ssi) {
+	ssidTagMapping[uid] = tag;
+}
+
+void DatabaseContext::getLatestCommitVersions(const Reference<LocationInfo>& locationInfo,
```
add comments describing the function

sbodagala replied:
> add comments describing the function (here, elsewhere)
Done already. The comments are in DatabaseContext.h. I can copy them here if you think that would be helpful.

> useful: `TraceEvent`s (commented out, but available for debugging)

dlambrig commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -169,6 +169,34 @@ void DatabaseContext::removeTssMapping(StorageServerInterface const& ssi) {
+	std::map<Version, std::set<Tag>> versionMap; // order the versions to be returned
+	for (int i = 0; i < locationInfo->locations()->size(); i++) {
+		UID uid = locationInfo->locations()->getId(i);
+		if (ssidTagMapping.find(uid) != ssidTagMapping.end()) {
```
is it expected to not be able to find the `uid` in `ssidTagMapping`

sbodagala replied:
> is it expected to not be able to find the `uid` in `ssidTagMapping` or would that be an error?

I don't know. If the proxy always has the most up-to-date UID-> Tag mapping then we should always find an entry here. Are there any exceptions to this on the proxy? Maybe Jingyu or Marcus can answer this question.

dlambrig commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -169,6 +169,34 @@ void DatabaseContext::removeTssMapping(StorageServerInterface const& ssi) {
+		UID uid = locationInfo->locations()->getId(i);
+		if (ssidTagMapping.find(uid) != ssidTagMapping.end()) {
+			Tag tag = ssidTagMapping[uid];
+			if (ssVersionVectorCache.hasVersion(tag)) {
```
it seems unexpected if we cannot find the tag in the array. If the SS has been written to it would have a valid version entry. Otherwise the entry would be `invalidVersion`.

dlambrig replied:
others [here](https://github.com/apple/foundationdb/pull/5182).

dlambrig commented on fdbclient/CommitProxyInterface.h:
```diff
@@ -293,9 +293,13 @@ struct GetKeyServerLocationsReply {
 	// if any storage servers in results have a TSS pair, that mapping is in here
 	std::vector<std::pair<UID, StorageServerInterface>> resultsTssMapping;
 
+	// maps storage server interfaces (captured in "results") to the tags of
```
document why we need logic to translate between ssi and tag

sfc-gh-mpilman replied:
Instead of adding this, wouldn't it be much easier to change the `getKeyServersLocations` response to also include the tag?

sbodagala replied:
> Instead of adding this, wouldn't it be much easier to change the `getKeyServersLocations` response to also include the tag?

You mean changing "results" in GetKeyServerLocationsReply from std::vector<std::pair<KeyRangeRef, vector<StorageServerInterface>>> to std::vector<std::pair<KeyRangeRef, vector<std::pair<StorageServerInterface, Tag>>>> ?

sfc-gh-mpilman replied:
Or maybe `Tag` could just be a member of `StorageServerInterface`? The tag should never change during the lifetime of that interface

sbodagala replied:
I like this idea.

dlambrig reviewed (changes requested):
add a few comments on design decisions and how tested
integrate/run with pr5114

sfc-gh-mpilman commented on fdbclient/DatabaseContext.h:
```diff
@@ -329,6 +329,9 @@ class DatabaseContext : public ReferenceCounted<DatabaseContext>, public FastAll
 	std::unordered_map<UID, Reference<TSSMetrics>> tssMetrics;
 
+	// map from ssid -> ss tag
+	std::unordered_map<UID, Tag> ssidTagMapping;
```
I think the code would be simpler if you instead change the `locationCache` to include the tag. I understand this might not be trivial, but it would simplify a lot of the code in `NativeAPI`

sbodagala replied:
I am not sure if it is logical to include the tag info in locationCache. Locationcache is mapping key range to locationInfo so not sure if the corresponding ssid -> tag map should be a part of that.

sfc-gh-mpilman replied:
`LocationInfo` is just a list of `StorageServerInterface` instances. Each of them are associated with one tag.

sfc-gh-mpilman replied:
I just realized: one difficulty here will be to pass the correct version to `loadBalance` when you send the request. I think a simple `std::min` would do the trick, but my guess is that it could potentially lead to more `transaction_too_old` errors. So a better solution would be to plumb this somehow into `loadBalance` (which is not exactly a pretty piece of code).

sbodagala replied:
I currently took the approach of including the latest commit versions of all storage server replicas (that serve the key) in the client request - this avoids changing the load balancer code and the receiver will still find its latest commit version in the request but at the cost of including "replication_factor" number of latest commit versions in the request (and probably higher if a split/merge is in progress), but probably this overhead is not that bad.

sfc-gh-mpilman replied:
I think that works as well -- so in practice you will ship between 3 and 6 tags to the storage, that doesn't sound like a meaningful overhead.

I still would prefer to get rid of the above structure though. If you store the tag in the `StorageServerInterface` you can just fetch the tags from your cache and keeping it up to date will be free (for your implementation)

sbodagala replied:
Here is my attempt at populating store tag in StorageServerInterface class:

```
diff --git a/fdbclient/StorageServerInterface.h b/fdbclient/StorageServerInterface.h
index bddca6fcb..9c5fe5496 100644
--- a/fdbclient/StorageServerInterface.h
+++ b/fdbclient/StorageServerInterface.h
@@ -58,6 +58,7 @@ struct StorageServerInterface {
        LocalityData locality;
        UID uniqueID;
        Optional<UID> tssPairID;
+       Tag tag;
 
        RequestStream<struct GetValueRequest> getValue;
        RequestStream<struct GetKeyRequest> getKey;
diff --git a/fdbserver/storageserver.actor.cpp b/fdbserver/storageserver.actor.cpp
index fe964eb6c..26a9c638c 100644
--- a/fdbserver/storageserver.actor.cpp
+++ b/fdbserver/storageserver.actor.cpp
@@ -5148,7 +5148,7 @@ ACTOR Future<Void> memoryStoreRecover(IKeyValueStore* store, Reference<ClusterCo
 
 // for creating a new storage server
 ACTOR Future<Void> storageServer(IKeyValueStore* persistentData,
-                                 StorageServerInterface ssi,
+                                 StorageServerInterface& ssi,
                                  Tag seedTag,
                                  Version tssSeedVersion,
                                  ReplyPromise<InitializeStorageReply> recruitReply,
@@ -5181,6 +5181,8 @@ ACTOR Future<Void> storageServer(IKeyValueStore* persistentData,
                        self.tag = seedTag;
                }
 
+               ssi.tag = self.tag;
+
                self.storage.makeNewStorageServerDurable();
                wait(self.storage.commit());
 
@@ -5208,7 +5210,7 @@ ACTOR Future<Void> storageServer(IKeyValueStore* persistentData,
        }
 }
 
-ACTOR Future<Void> replaceInterface(StorageServer* self, StorageServerInterface ssi) {
+ACTOR Future<Void> replaceInterface(StorageServer* self, StorageServerInterface& ssi) {
        ASSERT(!ssi.isTss());
        state Transaction tr(self->cx);
 
@@ -5271,6 +5273,7 @@ ACTOR Future<Void> replaceInterface(StorageServer* self, StorageServerInterface
                                                                self->tag = rep.tag;
                                                        }
                                                        self->allHistory = self->history;
+                                                       ssi.tag = self->tag;
 
                                                        TraceEvent("SSTag", self->thisServerID).detail("MyTag", self->tag.toString());
                                                        for (auto it : self->history) {
@@ -5299,7 +5302,7 @@ ACTOR Future<Void> replaceInterface(StorageServer* self, StorageServerInterface
        return Void();
 }
 
-ACTOR Future<Void> replaceTSSInterface(StorageServer* self, StorageServerInterface ssi) {
+ACTOR Future<Void> replaceTSSInterface(StorageServer* self, StorageServerInterface& ssi) {
        // RYW for KeyBackedMap
        state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(self->cx);
        state KeyBackedMap<UID, UID> tssMapDB = KeyBackedMap<UID, UID>(tssMappingKeys.begin);
@@ -5333,6 +5336,7 @@ ACTOR Future<Void> replaceTSSInterface(StorageServer* self, StorageServerInterfa
 
                        wait(tr->commit());
                        self->tag = myTag;
+                       ssi.tag = self->tag;
 
                        break;
                } catch (Error& e) {


```

sbodagala replied:
However, I am nervous about making this change (I am not even confident that this change is all that we need to do this). Maybe I will keep the code that I have for now and revisit this later?

sfc-gh-mpilman replied:
This looks correct to me. There should only be these two places where this is getting written. I would expect this approach to be less error prone than your current one.

sbodagala replied:
Thanks for going through the diff! I will apply this diff (and drop the ssid -> tag map) at some later point.

sbodagala commented:
NOTE: Added a new change set, extracted from https://github.com/apple/foundationdb/pull/5058 (the version indexer PR), to this PR. This change set makes a peek request block on a tlog until the tlog has a commit that is relevant to the requester.

jzhou77 commented on fdbserver/CommitProxyServer.actor.cpp:
```diff
@@ -1445,6 +1445,17 @@ void maybeAddTssMapping(GetKeyServerLocationsReply& reply,
+	for (const auto& [_, shard] : reply.results) {
+		for (auto& ssi : shard) {
+			auto iter = commitData->storageCache.find(ssi.id());
+			if (iter != commitData->storageCache.end()) {
```
I suppose this should always be true. In that case, we could `ASSERT_WE_THINK(iter != commitData->storageCache.end())` instead.

sbodagala replied:
Done.

sbodagala commented:
NOTE: Added a new change set that makes processes exchange version vector deltas.

sbodagala commented on fdbserver/CommitProxyServer.actor.cpp:
```diff
@@ -1449,9 +1449,8 @@ void addTagMapping(GetKeyServerLocationsReply& reply, ProxyCommitData* commitDat
-			if (iter != commitData->storageCache.end()) {
-				reply.resultsTagMapping.emplace_back(ssi.id(), iter->second->tag);
-			}
+			ASSERT_WE_THINK(iter != commitData->storageCache.end());
```
I was addressing Jingyu's review comment here. This is not related to the propagation of version vector deltas. Sorry for mixing these in the same commit.

sbodagala commented on fdbserver/GrvProxyServer.actor.cpp:
```diff
@@ -611,6 +612,7 @@ ACTOR Future<Void> sendGrvReplies(Future<GetReadVersionReply> replyFuture,
 		}
 		reply.midShardSize = midShardSize;
 		reply.tagThrottleInfo.clear();
+		grvProxyData->ssVersionVectorCache.getDelta(request.maxVersion, reply.ssVersionVectorDelta);
```
Shall I make getDelta() return the version vector? I chose to pass-by-reference to avoid an extra version vector copy. Let me know if the following version is going to be easier to read?

reply.ssVersionVectorDelta = grvProxyData->ssVersionVectorCache.getDelta(request.maxVersion);

dlambrig replied:
I like it how it is. faster, and not difficult to read.

sbodagala commented on fdbserver/CommitProxyServer.actor.cpp:
```diff
@@ -1132,7 +1132,7 @@ ACTOR Future<Void> postResolution(CommitBatchContext* self) {
 				when(wait(pProxyCommitData->cx->onProxiesChanged())) {}
 				when(GetRawCommittedVersionReply v = wait(pProxyCommitData->master.getLiveCommittedVersion.getReply(
-				         GetRawCommittedVersionRequest(waitVersionSpan.context, debugID),
+				         GetRawCommittedVersionRequest(waitVersionSpan.context, debugID, invalidVersion),
```
The sequencer will send the entire version vector  when replying to this request. Shall I add code to treat this as a special case and not send the version vector in the reply?

dlambrig replied:
I would add `// SOMEDAY: Special case this..etc`. And consider it when merging this project to trunk.

sbodagala replied:
Done.

jzhou77 commented on fdbclient/VersionVector.h:
```diff
@@ -69,6 +73,46 @@ struct VersionVector {
+		}
+
+		for (auto& iter : versionMap) {
+			delta.setVersion(iter.second, iter.first);
```
I see `maxVersion` is not set in `setVersion(const std::set<Tag>& tags, Version version)`, which is correct for the logic here. But is that intended? Then the `delta` doesn't have `maxVersion` set.

sbodagala replied:
Good find, we need to set maxVersion in that method too. (I missed setting maxVersion in that method when I pulled it from Dan's PR, commit a107dd655e04ff69ea5369aefa8f825c062a4f99).

The assertion failures in simulation tests, which I haven't investigated yet, are probably a result of not setting maxVersion in this method.

jzhou77 replied:
Note if you set `maxVersion` in that method, then next `setVersion` call would cause assertion failure.

sbodagala replied:
Sorry, why? Note we are ordering the entries (by version) before calling setVersion().

jzhou77 replied:
You are right. I missed the part that `versionMap` is ordered.

jzhou77 commented on fdbclient/VersionVector.h:
```diff
@@ -69,6 +73,46 @@ struct VersionVector {
+	}
+
+	void applyDelta(const VersionVector& delta) {
+		if (delta.maxVersion == invalidVersion) {
```
Related to the `maxVersion` not been set in `getDelta()`.

sbodagala replied:
Yes.

jzhou77 commented on fdbclient/VersionVector.h:
```diff
@@ -69,6 +73,46 @@ struct VersionVector {
+		}
+
+		std::map<Version, std::set<Tag>> versionMap;
+		for (auto& iter : versions) {
```
Nit: you can write more readable code like:
```
for (const auto& [tag, version] : versions) {
```

sbodagala replied:
Ok, agreed.

sbodagala replied:
Done.

sbodagala commented:
Rebase done. Uploaded the rebased version.

dlambrig commented on fdbclient/VersionVector.h:
```diff
@@ -64,6 +69,51 @@ struct VersionVector {
+		versions.clear();
+		maxVersion = invalidVersion;
+	}
+
```
- documentation here would be good for readers unfamiliar with this strategy.
- later, we may want to have a way to select how the vector is encoded. `knob={none, strategy1,  strategy2,}` etc).

sbodagala replied:
Done.

sbodagala commented on fdbclient/VersionVector.h:
```diff
@@ -83,33 +86,38 @@ struct VersionVector {
 			return;
 		}
 
-		ASSERT(maxVersion < delta.maxVersion);
```
Simulation tests were triggering this assert. The assertion failure was because a client doesn't need to make GetReadVersion requests sequentially, so there could be an overlap in the deltas received for successive read version requests.
